File: homing-server/src/main/resources/homing/js/hue/captains/singapura/js/homing/server/CssHandles.js
// =============================================================================
// CssHandles — the CSS manager's affiliate: the class handles (RFC 0002-ext1).
//
// Per-class JS handles use ES6 classes for prototype sharing and clean variant
// exposition. Each pseudo-state has a dedicated subclass mirroring the Java
// side (HoverVariantOf / FocusVariantOf / ActiveVariantOf): one uniform value
// type for plain classes, one subclass per variant state.
//
// A handle is IDENTITY, not behaviour: `st_btn` is the string "st-btn", and
// the rules behind it are bound late by the cascade. That is what lets the
// manager swap every sheet on the page while every holder of a handle keeps
// holding the same object (RFC 0064).
//
// Split out of CssClassManager when the manager took on the dependency graph
// and the theme switch; nothing here changed in the move.
// =============================================================================

class CssClass {
    constructor(name) { this.name = name; }
    toString() { return this.name; }
}

// Dedicated variant classes — mirror Java's HoverVariantOf / FocusVariantOf /
// ActiveVariantOf. Each carries the kebab-name of the state-restricted CSS
// rule. Distinct types so consumers can introspect via `instanceof`.
class HoverVariant  extends CssClass { static pseudoState = ":hover";  }
class FocusVariant  extends CssClass { static pseudoState = ":focus";  }
class ActiveVariant extends CssClass { static pseudoState = ":active"; }

// Lookup: variant state name (as emitted by CssGroupContentProvider) →
// dedicated subclass. Unknown states fall back to plain CssClass — keeps the
// JS forward-compatible if a new VariantOf subtype is added on the Java side
// before the JS is updated.
const VARIANT_CLASSES = {
    hover:  HoverVariant,
    focus:  FocusVariant,
    active: ActiveVariant,
};

class CssUtility extends CssClass {
    /**
     * @param {string} name           kebab-case base class name
     * @param {Object<string,string>} variants  state → variant kebab-name (e.g. { hover: "hover-bg-accent" })
     *
     * Each variant is precomputed once as the appropriate subclass instance
     * (HoverVariant / FocusVariant / etc.) and exposed as a PROPERTY (no
     * parens, no string return). This keeps every class handle in the system
     * the same uniform value type — `cls instanceof CssClass` for all of them
     * — while preserving distinguishability via the dedicated subclasses.
     *
     * Use site: `cn(bg_accent, bg_accent.hover)` — property access, no parens.
     */
    constructor(name, variants) {
        super(name);
        const states = Object.keys(variants || {});
        for (const state of states) {
            const VariantClass = VARIANT_CLASSES[state] || CssClass;
            this[state] = new VariantClass(variants[state]);
        }
    }
}

File: homing-server/src/main/java/hue/captains/singapura/js/homing/server/CssHandles.java
package hue.captains.singapura.js.homing.server;

import hue.captains.singapura.js.homing.core.EsModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;

import java.util.List;

/**
 * The CSS manager's affiliate: the class handles (RFC 0002-ext1). A handle is
 * identity, not behaviour — {@code st_btn} is the string {@code "st-btn"} and
 * the rules behind it are bound late by the cascade, which is what lets the
 * manager swap every sheet on the page while every holder keeps its handle
 * (RFC 0064). Split out of {@link CssClassManager} when the manager took on
 * the dependency graph and the theme switch.
 *
 * <p>The two exported names are the JS classes; they shadow the core
 * interfaces of the same names only inside this declaration.</p>
 */
public record CssHandles() implements EsModule<CssHandles> {

    public static final CssHandles INSTANCE = new CssHandles();

    /** The uniform handle type — every class handle in the system is one. */
    public record CssClass() implements Exportable._Constant<CssHandles> {}

    /** A base with precomputed variant handles as properties. */
    public record CssUtility() implements Exportable._Constant<CssHandles> {}

    @Override
    public ImportsFor<CssHandles> imports() {
        return ImportsFor.noImports();
    }

    @Override
    public ExportsOf<CssHandles> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new CssClass(), new CssUtility()));
    }
}
